Corrected and improved the Prepare > Column: Text > Split dialog

N-thony opened the pull request:
Fixes #6889 
@africanmathsinitiative/developers @rdstern this is ready for review. Thanks.

rdstern reviewed (changes requested):
@N-thony sorry I was wrong, you do need the str_split_fixed, when you fix the number of splits.  Otherwise you get a single variable with a c(  ) when there is more than one split.
I notice when you use the option for any number of splits it always puts the result at the end - and ignores the position setting.  I don't know if there is anything you can do about this?  If not, then maybe the Position button should be disabled or hidden for that option?

N-thony commented:
> @N-thony sorry I was wrong, you do need the str_split_fixed, when you fix the number of splits. Otherwise you get a single variable with a c( ) when there is more than one split. I notice when you use the option for any number of splits it always puts the result at the end - and ignores the position setting. I don't know if there is anything you can do about this? If not, then maybe the Position button should be disabled or hidden for that option?

Done.

rdstern commented:
Great, the main one is done.  Just checking on the other point, namely when getting Maximum number of components they are all at the end of the data, and not just after the column used - though it says they will be when you press that button.  I assume the current code can't cope with not knowing how many variables to include.  Maybe you can fix that, or perhaps disable the Position button for this option?

N-thony commented:
> Great, the main one is done. Just checking on the other point, namely when getting Maximum number of components they are all at the end of the data, and not just after the column used - though it says they will be when you press that button. I assume the current code can't cope with not knowing how many variables to include. Maybe you can fix that, or perhaps disable the Position button for this option?

Resolved

rdstern commented:
@N-thony that's great, but I now detect another problem.  The world_bank_pop is a good dataset to use, because that has a variable called indicator with contents like `SP.URB.TOTL`.
So I tried first without the regex, and then I ticked regex.  With your code it never runs with regex.  Your command is still: 
`split <- stringr::str_split(string=indicator, pattern=stringr::fixed("."), n=Inf, simplify=TRUE)`

With regex ticked it should become simply:
`split <- stringr::str_split(string=indicator, pattern=".", n=Inf, simplify=TRUE)`

Try it, and you will see the difference.

I am a bit concerned that is might also still be incorrect in the Find/Replace options?

N-thony commented:
> world_bank_pop

@rdstern I tried on my side and it works fine
![image](https://user-images.githubusercontent.com/68591383/139681531-f6d72b3a-4bd2-49c6-b6ec-a81d6e2c4ab2.png)

N-thony commented:
> @N-thony that's great, but I now detect another problem. The world_bank_pop is a good dataset to use, because that has a variable called indicator with contents like `SP.URB.TOTL`. So I tried first without the regex, and then I ticked regex. With your code it never runs with regex. Your command is still: `split <- stringr::str_split(string=indicator, pattern=stringr::fixed("."), n=Inf, simplify=TRUE)`
> 
> With regex ticked it should become simply: `split <- stringr::str_split(string=indicator, pattern=".", n=Inf, simplify=TRUE)`
> 
> Try it, and you will see the difference.
> 
> I am a bit concerned that is might also still be incorrect in the Find/Replace options?

@rdstern I have corrected the regex code. Please take a look.

rdstern reviewed (approved):
@N-thony if @shadrackkibet is ok, then this could be merged.

N-thony commented:
@HawardKetoyoMsatsi over to you.

HawardKetoyoMsatsi reviewed (changes requested):
The code looks good. Maybe if you could reduce blank lines 15-17 of dlgSplitText.vb to one line.

N-thony commented:
> The code looks good. Maybe if you could reduce blank lines 15-17 of dlgSplitText.vb to one line.

Done

N-thony commented:
> > The code looks good. Maybe if you could reduce blank lines 15-17 of dlgSplitText.vb to one line.
> 
> Done

@HawardKetoyoMsatsi I have resolved your comment above.

shadrackkibet commented on instat/dlgSplitText.vb:
```diff
@@ -80,6 +77,12 @@ Public Class dlgSplitText
+        ucrChkIncludeRegularExpressions.SetDefaultState(False)
+
+        ucrChkAddKeyboard.SetText("Add keyboard")
+        ucrChkAddKeyboard.Enabled = False
```
Why manual checking?

shadrackkibet commented on instat/dlgSplitText.vb:
```diff
@@ -80,6 +77,12 @@ Public Class dlgSplitText
         ucrNudPieces.SetMinMax(2, Integer.MaxValue)
 
+        ucrChkIncludeRegularExpressions.SetText("Include Regular Expressions")
+        ucrChkIncludeRegularExpressions.SetDefaultState(False)
```
Do you really need this?

shadrackkibet commented on instat/dlgSplitText.vb:
```diff
@@ -92,9 +95,16 @@ Public Class dlgSplitText
         clsBinaryColumns = New RFunction
+        clsStringCollFunction = New RFunction
+
+        ucrChkIncludeRegularExpressions.Checked = False
```
Same case here.

N-thony commented:
@shadrackkibet I resolved your comment above. Please take a look.

shadrackkibet commented on instat/dlgSplitText.vb:
```diff
@@ -71,15 +72,16 @@ Public Class dlgSplitText
         dctPatternPairs.Add("Hyphen -", Chr(34) & "-" & Chr(34))
         dctPatternPairs.Add("Underscore _", Chr(34) & "_" & Chr(34))
         ucrInputPattern.SetItems(dctPatternPairs)
-        'ucrInputPattern.SetRDefault(Chr(34) & " " & Chr(34)) 'This is the default for clsTextComponents
```
Could you explain why you removed this? It is good practice to first understand why the code was added initially before removing it.

N-thony commented on instat/dlgSplitText.vb:
```diff
@@ -143,15 +156,44 @@ Public Class dlgSplitText
+    End Sub
+
+    Private Sub ChangeParametersValues()
+        Dim strPattern As String = Nothing
+        Select Case ucrInputPattern.GetText
+            Case "Space ( )"
+                strPattern = Chr(34) & " " & Chr(34)
+            Case "Period ."
+                strPattern = Chr(34) & "." & Chr(34)
+            Case "Comma ,"
+                strPattern = Chr(34) & "," & Chr(34)
+            Case "Colon :"
+                strPattern = Chr(34) & ":" & Chr(34)
+            Case "Semicolon ;"
+                strPattern = Chr(34) & ";" & Chr(34)
+            Case "Hyphen -"
+                strPattern = Chr(34) & "-" & Chr(34)
+            Case "Underscore _"
+                strPattern = Chr(34) & "_" & Chr(34)
+        End Select
```
This is because of the value coming from the input control e.g Comma , we just need the comma symbol. I don't see another way to do that if not like this. Otherwise, you can suggest to me how to do it simply.

shadrackkibet replied:
This is clearly a duplication of what you've under `InitialiseDialog()`.

N-thony replied:
@shadrackkibet any suggestion about this?

shadrackkibet commented on instat/dlgSplitText.vb:
```diff
@@ -202,7 +204,15 @@ Public Class dlgSplitText
 
-    Private Sub ucrAll_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrChkIncludeRegularExpressions.ControlValueChanged, ucrReceiverSplitTextColumn.ControlValueChanged, ucrInputPattern.ControlValueChanged, ucrPnlSplitText.ControlValueChanged
+    Private Sub ucrChkIncludeRegularExpressions_ControlConteChanged(ucrChangedControl As ucrCore) Handles ucrChkIncludeRegularExpressions.ControlContentsChanged
+        If ucrChkIncludeRegularExpressions.Checked Then
+            clsSplitDummyFunction.AddParameter("checked", "TRUE", iPosition:=0)
+        Else
+            clsSplitDummyFunction.RemoveParameterByName("checked")
+        End If
```
Simply `.SetParameter` for this control.

shadrackkibet commented on instat/dlgSplitText.vb:
```diff
@@ -71,15 +72,16 @@ Public Class dlgSplitText
 
+        'This is not yet implemented
         ucrChkAddKeyboard.SetText("Add keyboard")
         ucrChkAddKeyboard.Enabled = False
```
We encourage the hiding of controls that do not work.

shadrackkibet commented on instat/dlgSplitText.vb:
```diff
@@ -67,19 +65,26 @@ Public Class dlgSplitText
         ucrNudPieces.SetParameter(New RParameter("n", 2))
         ucrNudPieces.SetMinMax(2, Integer.MaxValue)
 
+        ucrChkIncludeRegularExpressions.SetText("Include Regular Expressions")
```
Do we need this checkbox when splitting into binary columns?

N-thony replied:
I had a look at the code for the available option under splitting into binary and the R code, I think we don't need it. That is why is hidden when going to the splitting option

shadrackkibet reviewed (changes requested):
Under splitting into binary columns I got the following R code 
`split <- questionr::multi.split(var=village, split.char="instat.RFunction")`

@N-thony I would suggest you take time to first thoroughly test things before you even invite others to test them. Some obvious issues like the one above should be easy to detect if you test things thoroughly.

N-thony commented:
@shadrackkibet hopefully this is okay now. Take a look!

N-thony commented:
@shadrackkibet hopefully this is okay now? Take a look.

N-thony commented:
> @shadrackkibet hopefully this is okay now? Take a look.

@shadrackkibet how about this?

lloyddewit commented on instat/dlgSplitText.vb:
```diff
@@ -143,15 +148,47 @@ Public Class dlgSplitText
+
+    Private Sub ChangeParametersValues()
+        Dim strPattern As String = Nothing
+        Select Case ucrInputPattern.GetText
+            Case "Space ( )"
+                strPattern = Chr(34) & " " & Chr(34)
+            Case "Period ."
+                strPattern = Chr(34) & "." & Chr(34)
+            Case "Comma ,"
+                strPattern = Chr(34) & "," & Chr(34)
+            Case "Colon :"
+                strPattern = Chr(34) & ":" & Chr(34)
+            Case "Semicolon ;"
+                strPattern = Chr(34) & ";" & Chr(34)
+            Case "Hyphen -"
+                strPattern = Chr(34) & "-" & Chr(34)
+            Case "Underscore _"
+                strPattern = Chr(34) & "_" & Chr(34)
+            Case Else
+                strPattern = Chr(34) & ucrInputPattern.GetText & Chr(34)
+        End Select
```
If you have a dictionary that defines the correct character for each description (e.g. ',' for 'Comma ,'), then you can replace this `Select` with a single line.

N-thony commented:
@lloyddewit I made the changes regarding your comment above. Please take a look.

lloyddewit commented on instat/dlgSplitText.vb:
```diff
@@ -143,15 +159,37 @@ Public Class dlgSplitText
+
+        If dctPatternPairs.ContainsKey(ucrInputPattern.GetText) Then
+            strPattern = dctPatternPairs(ucrInputPattern.GetText)
+        Else
```
Is the user allowed to enter patterns that are not in the pre-defined list?

N-thony replied:
@lloyddewit yes.

lloyddewit reviewed (commented):
Looks good, just one open question

lloyddewit commented:
@shadrackkibet Thank you for your earlier review. There is one open comment. Please could you check that it's resolved?
I already set your other comments to resolved, I hope that's OK.
Thanks

lloyddewit commented:
@rdstern There are changes since you last tested. Please could you test again? thanks

rdstern reviewed (changes requested):
@N-thony please try again with the tidyr world bank population.  There is a variable that splits nicely into 3 with dot as the separator.  But it also works with regular expression ticked, which it shouldn't!  It currently gives split <- stringr::str_split_fixed(string=indicator, pattern=stringr::fixed("."), n=3)
so I think it is still not permitting the regular expression?

N-thony commented:
> @N-thony please try again with the tidyr world bank population. There is a variable that splits nicely into 3 with dot as the separator. But it also works with regular expression ticked, which it shouldn't! It currently gives split <- stringr::str_split_fixed(string=indicator, pattern=stringr::fixed("."), n=3) so I think it is still not permitting the regular expression?

@rdstern this seems to be fixed now. Please take a look. Thanks.

rdstern reviewed (approved):
I am happy now and approving.  @lloyddewit there is just a tiny code change since you accepted before.

lloyddewit commented:
@shadrackkibet there's still an open comment from you above, so we also need your approval before merging. Please could you take a look? thanks
